### PR TITLE
Add a new command line switch to skip analyzer execution

### LINKED ISSUE #40926: MSBuild parameter -p:RunAnalyzers=false is ignored
**Version Used**: 
```shell
dotnet --info
.NET Core SDK(global.json 반영):
 Version:   3.1.100
 Commit:    cd82f021f4

런타임 환경:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.100\

Host (useful for support):
  Version: 3.1.0
  Commit:  65f04fb6db

.NET Core SDKs installed:
  3.1.100 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Steps to Reproduce**:

1. Make a project
```shell
dotnet new console --name test-dotnet-run-analyzers
```
2. Add a bunch of analyzers
```shell
dotnet add package Microsoft.CodeAnalysis.FxCopAnalyzers
dotnet add package StyleCop.Analyzers
```
3. Build with `-p:RunAnalyzers=false`
Korean characters are mixed in the build output but I think you can get the point.
```shell
dotnet build -p:RunAnalyzers=false
.NET Core용 Microsoft (R) Build Engine 버전 16.4.0+e901037fe
Copyright (C) Microsoft Corporation. All rights reserved.

  24.47 ms에서 C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj에 대한 복원을 완료했습니다.
Program.cs(3,11): warning CA1707: 네임스페이스 이름 'test_dotnet_run_analyzers'에서 밑줄을 제거하세요. [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
Program.cs(1,1): warning SA1200: Using directive should appear within a namespace declaration [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
Program.cs(3,11): warning SA1300: Element 'test_dotnet_run_analyzers' should begin with an uppercase letter [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]       Program.cs(5,11): warning SA1400: Element 'Program' should declare an access modifier [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
Program.cs(7,21): warning SA1400: Element 'Main' should declare an access modifier [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
Program.cs(9,31): warning CA1303: 'void Program.Main(string[] args)' 메서드가 리터럴 문자열을 'void Console.WriteLine(string value)' 호출의 'value' 매개 변수로 전달합니다. 대신 리소스 테이블에서 "Hello World!" 문자열을 가져오세요. [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
Program.cs(7,35): warning CA1801: Main 메서드의 args 매개 변수가 사용되지 않았습니다. 매개 변수를 제거하거나 메서드 본문에 사용하세요. [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
CSC : warning SA0001: XML comment analysis is disabled due to project configuration [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
  test-dotnet-run-analyzers -> C:\Users\user\workspace\test-dotnet-run-analyzers\bin\Debug\netcoreapp3.1\test-dotnet-run-analyzers.dll

빌드했습니다.

Program.cs(3,11): warning CA1707: 네임스페이스 이름 'test_dotnet_run_analyzers'에서 밑줄을 제거하세요. [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
Program.cs(1,1): warning SA1633: The file header is missing or not located at the top of the file. [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
Program.cs(1,1): warning SA1200: Using directive should appear within a namespace declaration [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
Program.cs(3,11): warning SA1300: Element 'test_dotnet_run_analyzers' should begin with an uppercase letter [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]       Program.cs(5,11): warning SA1400: Element 'Program' should declare an access modifier [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
Program.cs(7,21): warning SA1400: Element 'Main' should declare an access modifier [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
Program.cs(9,31): warning CA1303: 'void Program.Main(string[] args)' 메서드가 리터럴 문자열을 'void Console.WriteLine(string value)' 호출의 'value' 매개 변수로 전달합니다. 대신 리소스 테이블에서 "Hello World!" 문자열을 가져오세요. [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
Program.cs(7,35): warning CA1801: Main 메서드의 args 매개 변수가 사용되지 않았습니다. 매개 변수를 제거하거나 메서드 본문에 사용하세요. [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
CSC : warning SA0001: XML comment analysis is disabled due to project configuration [C:\Users\user\workspace\test-dotnet-run-analyzers\test-dotnet-run-analyzers.csproj]
    경고 9개

경과 시간: 00:00:00.90
```
4. <PropertyGroup><RunAnalyzers>false</RunAnalyzers></PropertyGroup> doesn't work too.

**Expected Behavior**:
Warnings should not be printed

**Actual Behavior**:
Warnings are printed. Even it can make MSBuild to return an error with `-p:TreatWarningsAsErrors=true`

Also, it seems that the build warnings are not cached. If I build twice, there are no warnings in the second build. The recent release of Rust's cargo (https://github.com/rust-lang/cargo/blob/master/CHANGELOG.md#added-2) added a feature that caches build warnings. (https://github.com/rust-lang/cargo/pull/7450). I prefer to separate the build step and the lint step for CI. `dotnet clean` between the steps won't be necessary if we cache warnings.

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #40926

This support was added to `Microsoft.CodeAnalysis.targets` (ships with VS) in VS2019 16.5:
https://docs.microsoft.com/en-us/visualstudio/code-quality/disable-code-analysis?view=vs-2019.

However, the MSBuild property is not respected from 'dotnet' builds where this targets file is not imported (see https://github.com/dotnet/roslyn/issues/40926#issuecomment-574506716 for details).

This change moves the skip analyzers logic down to `Microsoft.Managed.Core.targets` and core compiler layer so it is respected for all builds. As per offline discussions with Jared and Chris, we can no longer skip analyzer execution from MSBuild by removing the "Analyzer" items as these can include source generators, which should never be skipped (at least there are no current scenarios where we want to skip generators). Hence, we are adding a new command line compiler switch `/skipAnalyzers`, optionally followed by a `+` or `-`, to allow skipping analyzers.

I verified the end-to-end functionality works fine by locally building `Microsoft.Net.Compilers.Toolset.Package` NuGet package and referencing it in a project.

### CODE REVIEW

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 1]
```diff
@@ -9129,6 +9129,66 @@ public void ReportAnalyzerOutput()
             CleanupAllGeneratedFiles(srcFile.Path);
         }
 
+        [Fact]
+        [WorkItem(40926, "https://github.com/dotnet/roslyn/issues/40926")]
+        public void SkipAnalyzersParse()
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:9134]: Feel like there are two tests missing:

1. Verify that even with `/skipAnalyzers-` the assembly is still loaded into the process. 
1. Verify that when a DLL has both an analyzer and a source generator that we load the source generator when `/skipAnalyzers-` is done. 

Honestly the second can really be a good proxy for the first.

File: src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets [hunk 2]
```diff
@@ -50,6 +50,28 @@
...
+        If 'RunAnalyzers' is not set, then fallback to user facing property 'RunAnalyzersDuringBuild'.
+        Latter property allows users to disable analyzers only for non-design time builds.
+        Design time builds are background builds inside Visual Studio,
+        see details here: https://github.com/dotnet/project-system/blob/master/docs/design-time-builds.md.
+        Setting 'RunAnalyzersDuringBuild' to false, without setting 'RunAnalyzers', allows users to
+        continue running analyzers in the background in Visual Studio while typing (i.e. intellisense),
+        while disabling analyzer execution on explicitly invoked non-design time builds.
+      -->
+    <PropertyGroup Condition="'$(RunAnalyzers)' == 'false' or
+                              ('$(RunAnalyzers)' == '' and '$(RunAnalyzersDuringBuild)' == 'false')">
```
Comment (jaredpar) [inline src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets:70]: Who, or what, would be expected to define `$(RunAnalyzersDuringBuild)`? Is that a property that we are setting today in the tasks / targets that we own?

Comment (mavasani) [reply src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets:70]: This is set by the project system when user clicks on the "Run on build" checkbox here: https://docs.microsoft.com/en-us/visualstudio/code-quality/disable-code-analysis?view=vs-2019#net-core-and-net-standard-projects

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs (hunk 1 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:9134]: Sure, I will add both tests.

File: src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets (hunk 2 above)
Comment (mavasani) [reply src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets:70]: More specifically: https://github.com/dotnet/project-system/blob/ad6afdd4f59e434f7624127fa5a6993d48641e15/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/GeneralBrowseObject.xaml#L412-L420

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs (hunk 1 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:9134]: Added a new test `TestSourceGeneratorsWithAnalyzers` which should test all combinations.

**Comment by chsienki** [comment]
@chsienki is added to the review. #Closed

File: src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets [hunk 3]
```diff
@@ -50,6 +50,28 @@
     </PropertyGroup>
   </Target>
 
+  <Target Name="_ComputeSkipAnalyzers" BeforeTargets="CoreCompile">
+    <!-- First, force clear non-user facing property '_SkipAnalyzers'. -->
```
Comment (jaredpar) [inline src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets:54]: Trust no one ;)

image

**Comment by mavasani** [comment]
Thanks for the quick reviews!

**Pull request merged**